### ISSUE TITLE
Error applying migration: BLOB, TEXT, GEOMETRY or JSON can't have a default value

### ISSUE THREAD

**Issue by OrihuelaConde** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#615]
### Steps to reproduce
I changed the name of a property and created a new migration:

```
migrationBuilder.RenameColumn(
                name: "ViajeSolicitante",
                table: "Formularios",
                newName: "ViajeSolicitanteNombre");
```

When I try to update the database, the next error occur.

### The issue
MySql.Data.MySqlClient.MySqlException (0x80004005): BLOB, TEXT, GEOMETRY or JSON column 'ViajeSolicitanteNombre' can't have a default value 

Failed executing DbCommand (9ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
ALTER TABLE `Formularios` CHANGE `ViajeSolicitante` `ViajeSolicitanteNombre` TEXT NOT NULL DEFAULT N''

### Further technical details

MySQL version: 5.7.22
Operating system: Windows 10 x64 1803
Pomelo.EntityFrameworkCore.MySql version: 2.1.0-rc1-final

Other details about my project setup: 

Visual Studio 15.7.3
Microsoft.AspNetCore.App 2.1

**Comment by mguinness** [comment]
Please provide your model, but a TEXT field cannot have a default value in MySQL.

**Comment by OrihuelaConde** [comment]
```
[Required(ErrorMessage = "Este campo es obligatorio.")]
[Column(TypeName = "TEXT")]
[Display(Name = "Nombres del solicitante principal")]
public string ViajeSolicitanteNombre { get; set; }
```

**Comment by caleblloyd** [comment]
Appears to be a bug with renaming.  We should look into this for 2.1

**Comment by rcl4444** [comment]
Pomelo.EntityFrameworkCore.MySql version: 2.1.0-rc2-20180621144030
when I roll back the change
`            migrationBuilder.RenameColumn(
                name: "PTId",
                table: "SplitRule_ProductClass",
                newName: "StintMark");`
`            migrationBuilder.RenameColumn(
                name: "StintMark",
                table: "SplitRule_ProductClass",
                newName: "PTId");`
![image](https://user-images.githubusercontent.com/19640490/41894060-4f26bcf6-7950-11e8-9002-b14c1317d6a8.png)
@caleblloyd

**Comment by mguinness** [comment]
@rcl4444 Please create a new issue as it appears unrelated to OP issue.  Have your tried anonymousArguments parameter with `new { columnType = "int" })` as suggested?

**Comment by caleblloyd** [comment]
I have tried to recreate this but am unable to.  We are about to publish RC2.  If the problem persists, please post a project that recreates this and we will investigate further.

**Comment by frankyjuang** [comment]
It seems that the issue still exists. Reproduce with two migrations. 
1. Initial
```
    public class TestModel
    {
        public long Id { get; set; }
        [Required]
        public string OldName { get; set; }
    }
```
2. Rename
```
    public class TestModel
    {
        public long Id { get; set; }
        [Required]
        public string NewName { get; set; }
    }
```
[testEF.zip](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/files/2218736/testEF.zip)

**Comment by kkwpsv** [comment]
After add a property with **REQUIRED** in a entity, the same thing happened
it will generate Migration like this


>      migrationBuilder.AddColumn<string>(
>                 name: "ColumnName",
>                 table: "TableName",
>                 nullable: false,
>                 defaultValue: "");

But I didn't set **DEFAULT VALUE,**.

**Comment by jamsoft** [comment]
I've just encountered this error using 2.1.2.  I created a migration and when executed against the database I got the error:

```
"ALTER TABLE `Reviews` ADD `Visibility` longtext NOT NULL DEFAULT '';" (627a98df)
MySql.Data.MySqlClient.MySqlException (0x80004005): BLOB, TEXT, GEOMETRY or JSON column 'Visibility' can't have a default value ---> MySql.Data.MySqlClient.MySqlException (0x80004005): BLOB, TEXT, GEOMETRY or JSON column 'Visibility' can't have a default value
```

The migration contains:

```
public partial class UpdatedEntities : Migration
{
    protected override void Up(MigrationBuilder migrationBuilder)
    {
        migrationBuilder.AddColumn<string>(
            name: "Visibility",
            table: "Reviews",
            nullable: false,
            defaultValue: "");
    }
    protected override void Down(MigrationBuilder migrationBuilder)
    {
        migrationBuilder.DropColumn(
            name: "Visibility",
            table: "Reviews");
    }
}
```

Manually hacking the migration and removing `defaultValue: ""` allows the migrations to run without errors.  There was no default value applied to this in the data context BUT this was using a ValueConverstion for an Enum ( a new feature in EF Core 2.1) not sure if that is the culprit?

**Comment by mguinness** [comment]
Would appear to be an issue in `Generate` method of [MySqlMigrationsSqlGenerator](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/EFCore.MySql/Migrations/MySqlMigrationsSqlGenerator.cs) class.

**Comment by jkdmyrs** [comment]
I'm also experiencing this issue when using the built in EnumToStringConverter from Microsoft.EntityFrameworkCore.Storage.ValueConversion with MySQL.

**Comment by jkdmyrs** [comment]
Quick update... this appears to only happen after the initial migration. In my initial migration, all my EnumToString conversions worked fine and EF was able to successfully run the Update-Database command. Then, I added a new enum, created a new migration, and then the Update-Database command failed with this issue. If I delete all my migrations, create a new "Initial Migration", drop the schema, and then run Update-Database, the issue does not occur.

**Comment by jkdmyrs** [comment]
Another quick note - This happens with the MySql.Data.EntityFrameworkCore pacakge also.

**Comment by mguinness** [comment]
@jkdmyrs Are you referring to [MySQL Connector/NET](https://dev.mysql.com/doc/connector-net/en/connector-net-entityframework-core.html) from Oracle?  If so, I don't see that provider mentioned in the issue that you referenced.

**Comment by jkdmyrs** [comment]
@mguinness the issue I referenced is for the Pomelo provider, yes. However, I have switched to the MySQL Connector/NET from Oracle provider and am still experiencing the same issue.

**Comment by darktimes** [comment]
Just faced exactly the same problem.

**Comment by Eternal21** [comment]
Same problem using Pomelo.EntityFrameworkCore.MySql 2.2.0: 

```
migrationBuilder.RenameColumn(
   name: "Value",
   table: "event_values",
   newName: "Text");
```

My current workaround is to just hack the migration by hand:

`migrationBuilder.Sql("ALTER TABLE event_values RENAME COLUMN Value TO Text;");`

**Comment by emysa341** [comment]
Same issue here 
**Model:**
public class FileAbstraction
    {
        public Guid     FileId   { get; set; }
        public string   FileSubPath { get; set; }
        public FileType FileType { get; set; }
        public long     Length   { get; set; }
    }
**Migration:** 
public partial class pa11 : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.RenameColumn(
                name: "FilePath",
                table: "File",
                newName: "FileSubPath");
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.RenameColumn(
                name: "FileSubPath",
                table: "File",
                newName: "FilePath");
        }
    }

**Error:**
Failed executing DbCommand (3ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
ALTER TABLE `File` CHANGE `FilePath` `FileSubPath` longtext NOT NULL DEFAULT '';

**Comment by lauxjpn** [comment]
This is already fixed in `3.0.0`.
Also, when #932 is merged, we will generate database version specific statements.

**Comment by mooop12** [comment]
@lauxjpn Is it fixed in 3.0.0-rc3.final? Because I'm still getting this exception for a migration created with that version.

**Comment by lauxjpn** [comment]
@mooop12 Yes it should be fixed. I tested it 4 days ago with @emysa341's model and migration. I can post the code I used for testing, if you want to verify it (that is always a good idea).
Otherwise, please post us your model in question, it's definition and the generated SQL script, so we can take a look at it or try to replicate your case.

**Comment by AnassL** [comment]
Got the same error with 3.0.0-rc3.final when trying to create the database from the newly generated migration file.

> Failed executing DbCommand (5ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
> CREATE TABLE `AspNetUserRoles` (
>     `UserId` varchar(255) CHARACTER SET utf8mb4 NOT NULL,
>     `RoleId` varchar(255) CHARACTER SET utf8mb4 NOT NULL,
>     `Id` longtext CHARACTER SET utf8mb4 NULL DEFAULT 'newsequentialid()',
>     CONSTRAINT `PK_AspNetUserRoles` PRIMARY KEY (`UserId`, `RoleId`),
>     CONSTRAINT `FK_AspNetUserRoles_AspNetRoles_RoleId` FOREIGN KEY (`RoleId`) REFERENCES `AspNetRoles` (`Id`) ON DELETE CASCADE,
>     CONSTRAINT `FK_AspNetUserRoles_AspNetUsers_UserId` FOREIGN KEY (`UserId`) REFERENCES `AspNetUsers` (`Id`) ON DELETE CASCADE
> );
> MySql.Data.MySqlClient.MySqlException (0x80004005): BLOB, TEXT, GEOMETRY or JSON column 'Id' can't have a default value

**Comment by lauxjpn** [comment]
It looks to me as if you altered the model, because the default `AspNetUserRoles` table would only contain the `UserId` and `RoleId` columns and would generate code like the following:

```
CREATE TABLE `AspNetUserRoles` (
    `UserId` varchar(255) CHARACTER SET utf8mb4 NOT NULL,
    `RoleId` varchar(255) CHARACTER SET utf8mb4 NOT NULL,
    CONSTRAINT `PK_AspNetUserRoles` PRIMARY KEY (`UserId`, `RoleId`),
    CONSTRAINT `FK_AspNetUserRoles_AspNetRoles_RoleId` FOREIGN KEY (`RoleId`) REFERENCES `AspNetRoles` (`Id`) ON DELETE CASCADE,
    CONSTRAINT `FK_AspNetUserRoles_AspNetUsers_UserId` FOREIGN KEY (`UserId`) REFERENCES `AspNetUsers` (`Id`) ON DELETE CASCADE
);
```

So it appears to me, that you added the `Id` property to the model, am I correct?

Your model definition might be wrong in multiple way (can't say for sure, because you didn't post it). The fact that you did specify `newsequentialid()` as the default, hints at that you want the `Id` property to be of type `System.Guid`.
The `newsequentialid()` is a SQL Server specific function and not available in MySQL.
Also, as the error message already tells you, the MySQL type `text` cannot have a direct default value (though depending on the used MySQL version, it would be theoretically possible to use an expression). Type like `char` and `varchar` can however.
Pomelo supports all GuidFormats of MySqlConnector and uses `char(36)` as the default.

If this does not fix your problem, you really need to post your model alterations, your model definition alterations and your migration `Up()` method.

**Comment by AnassL** [comment]
@lauxjpn you're right, that was my fault, I posted it in a hurry without checking the model, really sorry.

**Comment by mooop12** [comment]
@lauxjpn This is still an issue even on 3.0 release, though I have narrowed it down to being caused by having .NET Core 3 Nullable reference types enabled. As a reference, it is enabled by putting this line in the .csproj:

```
<Nullable>enable</Nullable>
```

**Comment by lauxjpn** [comment]
@mooop12 Please post us a model, a model definition and a migration to reproduce this issue (or just a working sample project).

**Comment by mooop12** [comment]
Just create a basic model with only an Id in it, add a migration, add a string value to the model, then add another migration (the default value is only set on migrations that change an existing table). This results in a migration like this:

```
migrationBuilder.AddColumn<string>(
                name: "Name",
                table: "Foos",
                nullable: false,
                defaultValue: "");
```

The final model is:

```
public class Foo {
    [Key]
    public int Id { get; set; }
    public string Name { get; set; }
}
```

I do have a repro project if you really need it but there really is nothing out of the ordinary in it, other than the nullability enabled.

**Comment by lauxjpn** [comment]
That is expected. If you enable `Nullable` mode, all string values are not nullable anymore by default. So their default value changes from `null` to an empty string instead.

Therefore `defaultValue: ""` in your migration `Up()` method is correct.

The issue in your code is, that you do not specify the length of the `Name` property, so you force Pomelo to use the MySQL `text` type. The `text` type does not support direct default values though and therefore an exception is thrown.

To fix your code, add the length of the string property to your model definition:

```c#
public class Foo {
    [Key]
    public int Id { get; set; }

    [MaxLength(255)] // <-- specify length to make Pomelo use varchar(255) instead of text
    public string Name { get; set; }
}
```

---

If you *need* the store type to be `text`, then this is also possible with recent versions of MySQL with `defaultValueSql: "('')"`. We are likely to add support for this in `3.1.0`, because we already prepared the support for this feature in a previous commit.

**Comment by mooop12** [comment]
Since the strings are not nullable, then a value must always be provided by the user for saving in which case I do not think a default value is necessary.

> The issue in your code is, that you do not specify the length of the Name property, so you force Pomelo to use the MySQL text type. The text type does not support direct default values though and therefore an exception is thrown.

That, IMO, is not an issue with my code. Users of EF providers should not be expected to know database nuances such as this one. At the very least an exception should be thrown instead of creating a migration that will not be possible to apply.

**Comment by lauxjpn** [comment]
> Since the strings are not nullable, then a value must always be provided by the user for saving in which case I do not think a default value is necessary.

Theoretically that is true. I checked this against SQL Server and it produces the same migration code including the `defaultValue: ""` (though it is not an issue for SQL Server, because it does not have the same limitations regarding the `text` type).

Enabling `Nullable` mode is just making the `System.String` property a `[Required]` property. So the same behavior will also happen, if you make it a required property explicitly.

This behavior generally makes sense because you could mark a previously nullable property as required at a later point in time, in which case you might like the migration to update any null values in the column to an empty string instead of throwing.

It might also be helpful so ensure, that in case a `null` value is send to the column anyway, which is still possible for explicitly marked as required properties and also for those implicitly marked as required by the `Nullable` mode, a default value is set instead of an exception being thrown. (Though that is probably not the reasoning behind this behavior because then the default value would also be set in conjunction with `CreateTable()` which it isn't.)

Maybe someone from the EF Core team (like @ajcvickers) can shed some light on the real reason behind this.

Anyway for our case here, you are right that we should not generate default values for columns that use store types that do not support default values.

**Comment by ajcvickers** [comment]
@lauxjpn EF creates default values in the migration when columns are added since there are often existing rows in the database and these rows need to get a default value. That being said, I'm not sure why the empty string is used here, or why it is not changed on AlterColumn.

@roji @bricelam Any ideas?

**Comment by roji** [comment]
For the column addition case, I did a quick test to see how MySQL behaves when a non-nullable text column is added to a table with existing rows:

```sql
CREATE TABLE test (id INT, first_name TEXT); 
INSERT INTO test (id, first_name) VALUES (1, 'foo');
ALTER TABLE test ADD COLUMN last_name TEXT NOT NULL;
SELECT * FROM test where last_name = '';
```

The last_name column for the existing row has an empty string; in other words, MySQL seems to have an implicit `default ''` behavior even if the explicit instruction in the ALTER TABLE statement isn't allowed. Because of this, I think it may be OK to modify your MigrationsSqlGenerator to simply omit the DEFAULT clause on certain types of columns which don't support it. This means it would continue to get generated in migrations, but would never actually make it to the database. @bricelam/@ajcvickers can confirm if that sounds like a good idea.

As to why we have a defaultValue when renaming a column, I have no idea... In a simple Npgsql test-case (changing a property's mapped column name only) no defaultValue gets generated. @bricelam probably knows.

> It might also be helpful so ensure, that in case a null value is send to the column anyway, which is still possible for explicitly marked as required properties and also for those implicitly marked as required by the Nullable mode, a default value is set instead of an exception being thrown. (Though that is probably not the reasoning behind this behavior because then the default value would also be set in conjunction with CreateTable() which it isn't.)

I'm not aware of any database which does this. AFAIK if you send a null value for a non-nullable column, you should get a constraint violation error even if that column defines a default value. In other words, the default is used only if a value is not provided when inserting, and not when NULL is provided.

FWIW here's [an issue](https://bugs.mysql.com/bug.php?id=21532) opened with MySQL 12 years to allow default...!

**Comment by lauxjpn** [comment]
> The last_name column for the existing row has an empty string; in other words, MySQL seems to have an implicit default '' behavior even if the explicit instruction in the ALTER TABLE statement isn't allowed. Because of this, I think it may be OK to modify your MigrationsSqlGenerator to simply omit the DEFAULT clause on certain types of columns which don't support it.

@roji Good observation! It it backed up by [11.7 Data Type Default Values](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html):

>Data type specifications can have explicit or implicit default values.
>[...]
>**Handling of Implicit Defaults:**
>If a data type specification includes no explicit `DEFAULT` value, MySQL determines the default value as follows:
>
>If the column can take `NULL` as a value, the column is defined with an explicit `DEFAULT NULL` clause.
>[...]
>For data entry into a `NOT NULL` column that has no explicit `DEFAULT` clause, if an `INSERT` or `REPLACE` statement includes no value for the column, or an `UPDATE` statement sets the column to `NULL`, MySQL handles the column according to the SQL mode in effect at the time:
>
>- If strict SQL mode is enabled, an error occurs for transactional tables and the statement is rolled back. For nontransactional tables, an error occurs, but if this happens for the second or subsequent row of a multiple-row statement, the preceding rows will have been inserted.
>
>- If strict mode is not enabled, MySQL sets the column to the implicit default value for the column data type.

I implemented the current handling at the migration level by changing a line in the model differ (that we already had) to just not generate the default values in general for those store types that don't support them. See #987:

```c#
public class MySqlMigrationsModelDiffer : MigrationsModelDiffer
{
    public MySqlMigrationsModelDiffer(
        [NotNull] IRelationalTypeMappingSource typeMappingSource,
        [NotNull] IMigrationsAnnotationProvider migrationsAnnotations,
        [NotNull] IChangeDetector changeDetector,
        [NotNull] IUpdateAdapterFactory updateAdapterFactory,
        [NotNull] CommandBatchPreparerDependencies commandBatchPreparerDependencies)
        : base(
            typeMappingSource,
            migrationsAnnotations,
            changeDetector,
            updateAdapterFactory,
            commandBatchPreparerDependencies)
    {
    }


    protected override IEnumerable<MigrationOperation> Add(IProperty target, DiffContext diffContext, bool inline = false)
    {
        if (target.FindTypeMapping() is RelationalTypeMapping storeType)
        {
            var valueGenerationStrategy = MySqlValueGenerationStrategyCompatibility.GetValueGenerationStrategy(MigrationsAnnotations.For(target).ToArray());


            // Ensure that null will be set for the columns default value, if CURRENT_TIMESTAMP has been required,
            // or when the store type of the column does not support default values at all.
            inline = inline ||
                        (storeType.StoreTypeNameBase == "datetime" ||
                        storeType.StoreTypeNameBase == "timestamp") &&
                        (valueGenerationStrategy == MySqlValueGenerationStrategy.IdentityColumn ||
                        valueGenerationStrategy == MySqlValueGenerationStrategy.ComputedColumn) ||
                        storeType.StoreTypeNameBase.Contains("text") ||
                        storeType.StoreTypeNameBase.Contains("blob") ||
                        storeType.StoreTypeNameBase == "geometry" ||
                        storeType.StoreTypeNameBase == "json";
        }


        return base.Add(target, diffContext, inline);
    }
}
```

Implementing it in a way for empty strings as you proposed is however a good backward compatible addition to what we have planed in an upcoming release, where we will natively support *default value expressions*, that are also allowed for the store types in question and would look like `DEFAULT ('')` instead of `DEFAULT ''` for empty strings, but are only supported by MySQL 18.0.13+.

> As to why we have a defaultValue when renaming a column, I have no idea... In a simple Npgsql test-case (changing a property's mapped column name only) no defaultValue gets generated. @bricelam probably knows.

The `RenameColumn` issue has already been solved. This #615 issue was recently reopend only because of the similar exception of `AddColumn`.

So the current question here is not about `RenameColumn`, but is the following:

Why is an empty default value generated for a required string property for `AddColumn`, while it is not generated for `AlterColumn` when an existing non-required string property is made required by the user?

(This is internal EF Core behavior that should be reproducible across all database providers. I tested it with Pomelo and SQL Server.)

Its this inconsistent behavior, that makes me suspicious about the `AddColumn` generation in the first place (or it could just mean, that is *has* a purpose that was just forgotten to be implemented for `AlterColumn`).

**Comment by roji** [comment]
> I already implemented a fix at the migration level by changing a line in the model differ (that we already had) to just not generate the default values in general for those store types that don't support them.

This is also a valid approach, but note that the model differ is an internal service (unlike MigrationsSqlGenerator) and ideally should not be touched. It may also make more "logical" sense to omit this at the SQL level: we really are using that new default, it's just that we can't use SQL to express it in MySQL. It would also work better in case where older migrations (with the default already scaffolded) need to run, or even if migration code has to work across providers. @bricelam will probably have some thoughts.

> Why is an empty default value generated for a required string property for AddColumn, while it is not generated for AlterColumn when an existing non-required string property is made required by the user?

At least in PostgreSQL I can confirm that it wouldn't help in any way: making a column non-nullable simply fails if there are any null values for that column, regardless of whether it has a default value or not:

```sql
CREATE TABLE data (id INT PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY, age INT NULL);
INSERT INTO data (age) VALUES (8), (NULL);
SELECT * FROM data;
ALTER TABLE data ALTER COLUMN age SET NOT NULL; -- ERROR: column "age" contains null values
ALTER TABLE data ALTER COLUMN age SET DEFAULT (3);
ALTER TABLE data ALTER COLUMN age SET NOT NULL; -- Same, ERROR: column "age" contains null values
```

We should check for SQL Server and MySQL, but my assumption is that this is somewhat "standard" behavior.

> Though I am not completely convinced that this should be handled by EF Core, because it is user database specific implementation behavior and should probably be implemented by the user, because he might not want an empty string but a different value as the replacement for NULL

I *think* users still have the option of defining a default value themselves on the new column being added if they want something other than an empty string (but I haven't tested this).

**Comment by ajcvickers** [comment]
@lauxjpn Are there places where this provider is using EF internal code? If so, we should follow up on that.

**Comment by lauxjpn** [comment]
> This is also a valid approach, but note that the model differ is an internal service (unlike MigrationsSqlGenerator) and ideally should not be touched.

That is true. We are aware of that (see #870 where we initially derived from it) and it is a bit ugly, but I considered the risk in this case rather low, because we really just alter the call to the base class method and don't really reimplement its functionality in any way.

The reason to implement it there was, that we had complaints from users in the past, where unneeded migration operations were added (sometimes only as a side effect).

But I reevalute this behavior again for .NET 5.

> I think users still have the option of defining a default value themselves on the new column being added if they want something other than an empty string (but I haven't tested this).

I agree (after I had some time to think about it a bit more now). It makes sense to generate the default `defaultValue` for `AlterColumn` and preset it to an empty string when a property is changed from non-required to required. It makes sense because the user is made aware that this might be an issue so he can adjust it as needed.

**Comment by lauxjpn** [comment]
@ajcvickers I think there are currently 2 places, where we override internal behavior, with this being one of them.

The other one is our `MySqlMigrator` class. We briefly talked about that one during our call but I have not yet opened an issue about it I said I would do.

See #939 for more details (it's way uglier than our `MySqlMigrationsModelDiffer` implementation so this is one we would really get rid of).

**Comment by ajcvickers** [comment]
@lauxjpn Thanks. I'll follow up later in the week.

**Comment by bricelam** [comment]
I agree with @roji--not generating SQL for the default value in MigrationsSqlGenerator would be better. But if you understand the risk of overriding the model differ, I'll let you decide.

One more thing to mention is that doing it in MigrationsSqlGenerator would also help users who manually add AddColumn assuming that they need to specify some default for new NOT NULL columns.

**Comment by lauxjpn** [comment]
> One more thing to mention is that doing it in MigrationsSqlGenerator would also help users who manually add AddColumn assuming that they need to specify some default for new NOT NULL columns.

I agree that this could be helpful for some users. But it also opens the window for subtle and unexpected bugs, where users now expect a default value to be applied to columns of store types that don't support them. With the current implementation, we at least openly throw an exception with a concise description about the fact that this is not supported.

**Comment by ilyamisura** [comment]
Maybe guys you forget to rename your DBSet in ApplicationDbContext

**Comment by bricelam** [comment]
@lauxjpn Sorry, late reply: I think this is a bug. I recently observed the same behavior and was confused too. Filed https://github.com/dotnet/efcore/issues/19882

**Comment by lauxjpn** [comment]
We need to check, that our issue got fixed with the fix for https://github.com/dotnet/efcore/issues/19882 upstream.

**Comment by foriequal0** [comment]
EF Core 6.0.5 generates following migration when the model is changed from
```c#
class Model
{
    public string Blah { get; set; }
    [Column(TypeName = "json")]
    public string Meh { get; set; }
}
```
into 
```c#
class Model
{
    public string Blah1 { get; set; }
    [Column(TypeName = "json")]
    public string Meh1 { get; set; }
}
```

```c#
migrationBuilder.RenameColumn(name: "Meh", table: "Model", newName: "Meh1");
migrationBuilder.RenameColumn(name: "Blah", table: "Model", newName: "Blah1");
```

Then Pomelo.EntityFrameworkCore.MySql 6.0.1 generates following script:
```sql
ALTER TABLE `Model` CHANGE `Meh` `Meh1` json NOT NULL;

ALTER TABLE `Model` CHANGE `Blah` `Blah1` varchar(255) NOT NULL DEFAULT '';
```

So, it is fixed...?

**Comment by lauxjpn** [comment]
@foriequal0 Yes, all issues here are fixed:

* The original one, regarding column renaming, had been fixed for Pomelo 3.0.0.
* The second one, regarding `AddColumn()` and default values, was fixed by #987.
* The last one, regarding inconsistent behavior between `AddColumn()` and `AlterColumn()` in conjunction with required properies, was fixed by https://github.com/dotnet/efcore/issues/19882.

It should be mentioned, that in Pomelo 6.0, we will only add default values when they are expected by the user. Making an existing string column _required_ in a later migration, will _not_ lead to a default value being added to the table definition. Instead, an `UPDATE` statement will be generated, that updates existing `NULL` values in the column to an empty string (see #1498).
